server/etcdserver/api/v3discovery/discovery_test.go

```diff
@@ -0,0 +1,651 @@
+func (fkv *fakeKVForClusterMembers) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	kvs := memberInfoToKeyValues(fkv.members)
+
+	return &clientv3.GetResponse{
+		Header: &etcdserverpb.ResponseHeader{
+			Revision: 10,
+		},
+		Kvs: kvs,
+	}, nil
+}
+
+func memberInfoToKeyValues(members []memberInfo) []*mvccpb.KeyValue {
+	kvs := make([]*mvccpb.KeyValue, 0)
+	for _, mi := range members {
+		kvs = append(kvs, &mvccpb.KeyValue{
+			Key:            []byte(mi.peerRegKey),
+			Value:          []byte(mi.peerURLsMap),
+			CreateRevision: mi.createRev,
+		})
+	}
+
+	return kvs
+}
+
+func TestGetClusterMembers(t *testing.T) {
+	lg, err := zap.NewProduction()
+	if err != nil {
+		t.Errorf("Failed to create a logger, error: %v", err)
+	}
+
+	d := &discovery{
+		lg: lg,
+		c: &clientv3.Client{
+			KV: &fakeKVForClusterMembers{
+				fakeBaseKV: &fakeBaseKV{},
+				members:    actualMemberInfo,
+			},
+		},
+		cfg:          &DiscoveryConfig{},
+		clusterToken: clusterFakeToken,
+	}
+
+	clsInfo, _, err := d.getClusterMembers()
+	if err != nil {
+		t.Errorf("Failed to get cluster members, error: %v", err)
+	}
+
+	if clsInfo.Len() != len(expectedMemberInfo) {
+		t.Errorf("unexpected member count, expected: %d, got: %d", len(expectedMemberInfo), clsInfo.Len())
+	}
+
+	for i, m := range clsInfo.members {
+		if m != expectedMemberInfo[i] {
+			t.Errorf("unexpected member[%d], expected: %v, got: %v", i, expectedMemberInfo[i], m)
+		}
+	}
+}
+
+// fakeKVForCheckCluster is used to test checkCluster.
+type fakeKVForCheckCluster struct {
+	*fakeBaseKV
+	t                 *testing.T
+	token             string
+	clusterSizeStr    string
+	members           []memberInfo
+	getSizeRetries    int
+	getMembersRetries int
+}
+
+// We only need to overwrite method `Get`.
+func (fkv *fakeKVForCheckCluster) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	clusterSizeKey := fmt.Sprintf("/_etcd/registry/%s/_config/size", fkv.token)
+	clusterMembersKey := fmt.Sprintf("/_etcd/registry/%s/members", fkv.token)
+
+	if key == clusterSizeKey {
+		if fkv.getSizeRetries > 0 {
+			fkv.getSizeRetries--
+			// discovery client should retry on error.
+			return nil, errors.New("get cluster size failed")
+		}
+		return &clientv3.GetResponse{
+			Kvs: []*mvccpb.KeyValue{
+				{
+					Value: []byte(fkv.clusterSizeStr),
+				},
+			},
+		}, nil
+
+	} else if key == clusterMembersKey {
+		if fkv.getMembersRetries > 0 {
+			fkv.getMembersRetries--
+			// discovery client should retry on error.
+			return nil, errors.New("get cluster members failed")
+		}
+		kvs := memberInfoToKeyValues(fkv.members)
+
+		return &clientv3.GetResponse{
+			Header: &etcdserverpb.ResponseHeader{
+				Revision: 10,
+			},
+			Kvs: kvs,
+		}, nil
+	} else {
+		fkv.t.Errorf("unexpected key: %s", key)
+		return nil, fmt.Errorf("unexpected key: %s", key)
+	}
+}
+
+func TestCheckCluster(t *testing.T) {
+	cases := []struct {
```

Comment:
Please name each test case. Name should describe expected results (expectError in this case) and why. For example `Should succeed if member exist`. 

Name of the tests can be utilized in subtest by passing it to `t.Run(name, func{})`. This will make failures for each case separate and make nice test names that explain intention of the author. For example test name for case would be: `TestCheckCluster/Should_succeed_if_member_exist`

Reply:
Resolved & thanks.